File: SRC/ToolBox/ToolboxXmlManager.cs
using System;
using System.Windows.Forms;
using System.Xml;
using Microsoft.VisualBasic;

namespace ToolBox.ToolboxLibrary
{
    /// <summary>
    /// ToolboxXmlManager - Reads an XML file and populates the toolbox.
    /// </summary>
    public class ToolboxXmlManager
    {
        private Toolbox m_toolbox = null;
        private Type[] cpcdosFormsToolTypes = new Type[] { typeof(TextBlock), typeof(Button), typeof(CheckBox), typeof(PictureBox), typeof(ProgressBar), typeof(TextBox) };
        private ToolBox toolBox;

        // Private componentsToolTypes As Type() = New Type() {GetType(System.IO.FileSystemWatcher), GetType(System.Diagnostics.Process), GetType(Timers.Timer)}
        // Private dataToolTypes As Type() = New Type() {GetType(Data.OleDb.OleDbCommandBuilder), GetType(Data.OleDb.OleDbConnection), GetType(Data.SqlClient.SqlCommandBuilder), GetType(Data.SqlClient.SqlConnection)}
        // Private userControlsToolTypes As Type() = New Type() {GetType(UserControl)}

        public ToolboxXmlManager(Toolbox toolbox)
        {
            m_toolbox = toolbox;
        }

        public ToolboxXmlManager(ToolBox toolBox)
        {
            this.toolBox = toolBox;
        }

        public ToolboxTabCollection PopulateToolboxInfo()
        {
            try
            {
                if (object.Equals(Toolbox.FilePath, null) || object.Equals(Toolbox.FilePath, "") || object.Equals(Toolbox.FilePath, string.Empty))
                    return PopulateToolboxTabs();
                var xmlDocument = new XmlDocument();
                xmlDocument.Load(Toolbox.FilePath);
                return PopulateToolboxTabs(xmlDocument);
            }
            catch (Exception ex)
            {
                MessageBox.Show("Error occured in reading Toolbox.xml file." + Constants.vbLf + ex.ToString());
                return null;
            }
        }

        private Toolbox Toolbox
        {
            get
            {
                return m_toolbox;
            }
        }

        private ToolboxTabCollection PopulateToolboxTabs()
        {
            var toolboxTabs = new ToolboxTabCollection();
            var tabNames = new[] { Strings.CpcdosForms }; // Strings.Components, Strings.Data, Strings.UserControls}
            for (int i = 0, loopTo = tabNames.Length - 1; i <= loopTo; i++)
            {
                var toolboxTab = new ToolboxTab();
                toolboxTab.Name = tabNames[i];
                PopulateToolboxItems(toolboxTab);
                toolboxTabs.Add(toolboxTab);
            }

            return toolboxTabs;
        }

        private void PopulateToolboxItems(ToolboxTab toolboxTab)
        {
            if (toolboxTab is null)
                return;
            Type[] typeArray = null;
            switch (toolboxTab.Name ?? "")
            {
                case Strings.CpcdosForms:
                    {
                        typeArray = cpcdosFormsToolTypes;
                        break;
                    }
                    // Case Strings.Components
                    // typeArray = componentsToolTypes
                    // Case Strings.Data
                    // typeArray = dataToolTypes
                    // Case Strings.UserControls
                    // typeArray = userControlsToolTypes
                    // Case Else
            }

            var toolboxItems = new ToolboxItemCollection();
            for (int i = 0, loopTo = typeArray.Length - 1; i <= loopTo; i++)
            {
                var toolboxItem = new ToolboxItem();
                toolboxItem.Type = typeArray[i];
                toolboxItem.Name = typeArray[i].Name;
                toolboxItems.Add(toolboxItem);
            }

            toolboxTab.ToolboxItems = toolboxItems;
        }

        private ToolboxTabCollection PopulateToolboxTabs(XmlDocument xmlDocument)
        {
            if (xmlDocument is null)
                return null;
            var toolboxNode = xmlDocument.FirstChild;
            if (toolboxNode is null)
                return null;
            var tabCollectionNode = toolboxNode.FirstChild;
            if (tabCollectionNode is null)
                return null;
            var tabsNodeList = tabCollectionNode.ChildNodes;
            if (tabsNodeList is null)
                return null;
            var toolboxTabs = new ToolboxTabCollection();
            foreach (XmlNode tabNode in tabsNodeList)
            {
                if (tabNode is null)
                    continue;
                var propertiesNode = tabNode.FirstChild;
                if (propertiesNode is null)
                    continue;
                XmlNode nameNode = propertiesNode[Strings.Name];
                if (nameNode is null)
                    continue;
                var toolboxTab = new ToolboxTab();
                toolboxTab.Name = nameNode.InnerXml.ToString();
                PopulateToolboxItems(tabNode, toolboxTab);
                toolboxTabs.Add(toolboxTab);
            }

            if (toolboxTabs.Count == 0)
                return null;
            return toolboxTabs;
        }

        private void PopulateToolboxItems(XmlNode tabNode, ToolboxTab toolboxTab)
        {
            if (tabNode is null)
                return;
            XmlNode toolboxItemCollectionNode = tabNode[Strings.ToolboxItemCollection];
            if (toolboxItemCollectionNode is null)
                return;
            var toolboxItemNodeList = toolboxItemCollectionNode.ChildNodes;
            if (toolboxItemNodeList is null)
                return;
            var toolboxItems = new ToolboxItemCollection();
            foreach (XmlNode toolboxItemNode in toolboxItemNodeList)
            {
                if (toolboxItemNode is null)
                    continue;
                XmlNode typeNode = toolboxItemNode[Strings.Type];
                if (typeNode is null)
                    continue;
                bool found = false;
                var loadedAssemblies = AppDomain.CurrentDomain.GetAssemblies();
                int i = 0;
                while (i < loadedAssemblies.Length && !found)
                {
                    var assembly = loadedAssemblies[i];
                    var types = assembly.GetTypes();
                    int j = 0;
                    while (j < types.Length && !found)
                    {
                        var type = types[j];
                        if (Equals(type.FullName, typeNode.InnerXml.ToString()))
                        {
                            var toolboxItem = new ToolboxItem();
                            toolboxItem.Type = type;
                            toolboxItems.Add(toolboxItem);
                            found = true;
                        }

                        j += 1;
                    }

                    i += 1;
                }
            }

            toolboxTab.ToolboxItems = toolboxItems;
            return;
        }

        private class Strings
        {
            public const string Toolbox = "Toolbox";
            public const string TabCollection = "TabCollection";
            public const string Tab = "Tab";
            public const string Properties = "Properties";
            public const string Name = "Name";
            public const string ToolboxItemCollection = "ToolboxItemCollection";
            public const string ToolboxItem = "ToolboxItem";
            public const string Type = "Type";
            public const string CpcdosForms = "Objets Cpcdos";
            public const string Components = "Components";
            // Public Const Data = "Data"
            public const string UserControls = "User Controls";
        }
    } // class
} // namespace

File: SRC/ToolBox/ToolboxXmlManager.vb
Imports System
Imports System.Xml
Imports System.Windows.Forms
Imports DevComponents.DotNetBar

Namespace ToolboxLibrary
    ''' <summary>
    ''' ToolboxXmlManager - Reads an XML file and populates the toolbox.
    ''' </summary>
    Public Class ToolboxXmlManager
        Private m_toolbox As Toolbox = Nothing
        Private cpcdosFormsToolTypes As Type() = New Type() {GetType(Button), GetType(CheckBox), GetType(TextBlock), GetType(TextBox), GetType(PictureBox), GetType(ProgressBar), GetType(Window), GetType(ListBox), GetType(Explorer)}
        'Private componentsToolTypes As Type() = New Type() {GetType(System.IO.FileSystemWatcher), GetType(System.Diagnostics.Process), GetType(Timers.Timer)}
        'Private dataToolTypes As Type() = New Type() {GetType(Data.OleDb.OleDbCommandBuilder), GetType(Data.OleDb.OleDbConnection), GetType(Data.SqlClient.SqlCommandBuilder), GetType(Data.SqlClient.SqlConnection)}
        'Private userControlsToolTypes As Type() = New Type() {GetType(UserControl)}

        Public Sub New(ByVal toolbox As Toolbox)
            m_toolbox = toolbox
        End Sub

        Public Function PopulateToolboxInfo() As ToolboxTabCollection
            Try
                If Equals(Toolbox.FilePath, Nothing) OrElse Equals(Toolbox.FilePath, "") OrElse Equals(Toolbox.FilePath, String.Empty) Then Return PopulateToolboxTabs()
                Dim xmlDocument As XmlDocument = New XmlDocument()
                xmlDocument.Load(Toolbox.FilePath)
                Return PopulateToolboxTabs(xmlDocument)
            Catch ex As Exception
                MessageBox.Show("Error occured in reading Toolbox.xml file." & Microsoft.VisualBasic.Constants.vbLf & ex.ToString())
                Return Nothing
            End Try
        End Function

        Private ReadOnly Property Toolbox As Toolbox
            Get
                Return m_toolbox
            End Get
        End Property

        Private Function PopulateToolboxTabs() As ToolboxTabCollection
            Dim toolboxTabs As ToolboxTabCollection = New ToolboxTabCollection()
            Dim tabNames = {Strings.CpcdosForms} 'Strings.Components, Strings.Data, Strings.UserControls}

            For i = 0 To tabNames.Length - 1
                Dim toolboxTab As ToolboxTab = New ToolboxTab()
                toolboxTab.Name = tabNames(i)
                PopulateToolboxItems(toolboxTab)
                toolboxTabs.Add(toolboxTab)
            Next

            Return toolboxTabs
        End Function

        Private Sub PopulateToolboxItems(ByVal toolboxTab As ToolboxTab)
            If toolboxTab Is Nothing Then Return
            Dim typeArray As Type() = Nothing

            Select Case toolboxTab.Name
                Case Strings.CpcdosForms
                    typeArray = cpcdosFormsToolTypes
                    '  Case Strings.Components
                    ' typeArray = componentsToolTypes
                    '  Case Strings.Data
                    ' typeArray = dataToolTypes
                    ' Case Strings.UserControls
                    '  typeArray = userControlsToolTypes
                    'Case Else
            End Select

            Dim toolboxItems As ToolboxItemCollection = New ToolboxItemCollection()

            For i = 0 To typeArray.Length - 1
                Dim toolboxItem As ToolboxItem = New ToolboxItem()
                toolboxItem.Type = typeArray(i)
                toolboxItem.Name = typeArray(i).Name
                toolboxItems.Add(toolboxItem)
            Next

            toolboxTab.ToolboxItems = toolboxItems
        End Sub

        Private Function PopulateToolboxTabs(ByVal xmlDocument As XmlDocument) As ToolboxTabCollection
            If xmlDocument Is Nothing Then Return Nothing
            Dim toolboxNode = xmlDocument.FirstChild
            If toolboxNode Is Nothing Then Return Nothing
            Dim tabCollectionNode = toolboxNode.FirstChild
            If tabCollectionNode Is Nothing Then Return Nothing
            Dim tabsNodeList = tabCollectionNode.ChildNodes
            If tabsNodeList Is Nothing Then Return Nothing
            Dim toolboxTabs As ToolboxTabCollection = New ToolboxTabCollection()

            For Each tabNode As XmlNode In tabsNodeList
                If tabNode Is Nothing Then Continue For
                Dim propertiesNode = tabNode.FirstChild
                If propertiesNode Is Nothing Then Continue For
                Dim nameNode As XmlNode = propertiesNode(Strings.Name)
                If nameNode Is Nothing Then Continue For
                Dim toolboxTab As ToolboxTab = New ToolboxTab()
                toolboxTab.Name = nameNode.InnerXml.ToString()
                PopulateToolboxItems(tabNode, toolboxTab)
                toolboxTabs.Add(toolboxTab)
            Next

            If toolboxTabs.Count = 0 Then Return Nothing
            Return toolboxTabs
        End Function

        Private Sub PopulateToolboxItems(ByVal tabNode As XmlNode, ByVal toolboxTab As ToolboxTab)
            If tabNode Is Nothing Then Return
            Dim toolboxItemCollectionNode As XmlNode = tabNode(Strings.ToolboxItemCollection)
            If toolboxItemCollectionNode Is Nothing Then Return
            Dim toolboxItemNodeList = toolboxItemCollectionNode.ChildNodes
            If toolboxItemNodeList Is Nothing Then Return
            Dim toolboxItems As ToolboxItemCollection = New ToolboxItemCollection()

            For Each toolboxItemNode As XmlNode In toolboxItemNodeList
                If toolboxItemNode Is Nothing Then Continue For
                Dim typeNode As XmlNode = toolboxItemNode(Strings.Type)
                If typeNode Is Nothing Then Continue For
                Dim found = False
                Dim loadedAssemblies As System.Reflection.Assembly() = AppDomain.CurrentDomain.GetAssemblies()
                Dim i = 0

                While i < loadedAssemblies.Length AndAlso Not found
                    Dim assembly = loadedAssemblies(i)
                    Dim types As Type() = assembly.GetTypes()
                    Dim j = 0

                    While j < types.Length AndAlso Not found
                        Dim type = types(j)

                        If Equals(type.FullName, typeNode.InnerXml.ToString()) Then
                            Dim toolboxItem As ToolboxItem = New ToolboxItem()
                            toolboxItem.Type = type
                            toolboxItems.Add(toolboxItem)
                            found = True
                        End If

                        j += 1
                    End While

                    i += 1
                End While
            Next

            toolboxTab.ToolboxItems = toolboxItems
            Return
        End Sub

        Private Class Strings
            Public Const Toolbox = "Toolbox"
            Public Const TabCollection = "TabCollection"
            Public Const Tab = "Tab"
            Public Const Properties = "Properties"
            Public Const Name = "Name"
            Public Const ToolboxItemCollection = "ToolboxItemCollection"
            Public Const ToolboxItem = "ToolboxItem"
            Public Const Type = "Type"
            Public Const CpcdosForms = "Objets Cpcdos"
            Public Const Components = "Components"
            'Public Const Data = "Data"
            Public Const UserControls = "User Controls"
        End Class
    End Class ' class
End Namespace ' namespace

